### PR TITLE
add plate element

### PR DESCRIPTION
**Pull request by obucklin**
This adds the plate element to Compas_timber. This is concieved as an extrusion with a boundary curve(Polyline) and a thickness. A vector input serves to switch the side towards which the plate is extruded. This has been tested with the Drill feature to ensure the Brep is oriented correctly and the boolean subtraction functions properly. Certain attributes are not yet implemented fully, such as the bounding box. TimberModel has been updated to include the `TimberModel.plates` property. 

I will continue to change other modules to remove references to `beam` in exchange for more generic `element` types when appropriate. 

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

invokes don't seem to be working on mac, I will run these when I get back to my PC.

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas_timber/elements/plate.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    # ==========================================================================
+    # Alternative constructors
+    # ==========================================================================
+
+    @staticmethod
+    def _create_shape(outline, vector):
+        brep = Brep.from_extrusion(outline, vector)
+
+        return brep
```
Comment (chenkasirer) [inline src/compas_timber/elements/plate.py:207]:
```suggestion
        return Brep.from_extrusion(outline, vector)
```

File: src/compas_timber/elements/plate.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        -------
+        :class:`compas.datastructures.Mesh`
+            The collision geometry of the element.
+
+        """
+        return self.compute_aabb.to_mesh()
+
+    # ==========================================================================
+    # Alternative constructors
+    # ==========================================================================
```
Comment (chenkasirer) [inline src/compas_timber/elements/plate.py:201]: move down or remove this section header as there are currently no alternative constructors

File: src/compas_timber/model/model.py [hunk 3]
```diff
@@ -45,18 +48,26 @@ def __from_data__(cls, data):
...
         super(TimberModel, self).__init__()
         self._beams = []
+        self._plates = []
         self._walls = []
         self._joints = []
         self._topologies = []  # added to avoid calculating multiple times
 
     def __str__(self):
-        return "TimberModel ({}) with {} beam(s) and {} joint(s).".format(self.guid, len(self.beams), len(self.joints))
+        return "TimberModel ({}) with {} beam(s), {} plate(s) and {} joint(s).".format(
```
Comment (chenkasirer) [inline src/compas_timber/model/model.py:57]: not necessarily related to this PR, but this is getting long. maybe we can format this similarly to `Brep`?
https://github.com/compas-dev/compas/blob/6e85276ce10d485e6b15f895b127c72a6862c7b4/src/compas/geometry/brep/brep.py#L153

File: src/compas_timber/ghpython/components/CT_Plate/code.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                    update_rhobj_attributes_name(guid, "category", c)
+
+                plates.append(plate)
+                scene.add(plate.shape)
+
+        geo = scene.draw()
+
+        return plates, geo
+
+    def _get_guid_and_geometry(self, line):
```
Comment (chenkasirer) [inline src/compas_timber/ghpython/components/CT_Plate/code.py:77]: could be in a future PR, but maybe we add this to the `ghpython_helpers` module

**Review by chenkasirer — CHANGES_REQUESTED** [review]
Generally looks good, added some comments on the code.
Two general question/requests:
- please add unittests for the new `Plate`
- have you had a look at https://github.com/BlockResearchGroup/compas_model/blob/main/src/compas_model/elements/plate.py? maybe it's not right for us but might be worth it to have a look.

**Comment by obucklin** [comment]
> * have you had a look at https://github.com/BlockResearchGroup/compas_model/blob/main/src/compas_model/elements/plate.py? maybe it's not right for us but might be worth it to have a look.

I had not seen that. the two polygons as faces seems like a good way to do it. I will try it out and see how it works. My only concern is whether their shape (a mesh) works with the CT features. Maybe you know...

File: src/compas_timber/elements/plate.py (hunk 1 above)
Comment (obucklin) [reply src/compas_timber/elements/plate.py:207]: done

File: src/compas_timber/elements/plate.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        super(Plate, self).__init__(**kwargs)
+        if not outline.is_closed:
+            raise ValueError("The outline points are not coplanar.")
+        self.outline = outline
+        self.thickness = thickness
+        self.features = []
+        self.attributes = {}
+        self.attributes.update(kwargs)
+        self.debug_info = []
+        self.frame = Frame.from_points(outline.points[0], outline.points[1], outline.points[-2])
```
[the remark this answers is not in the document]
Comment (obucklin) [reply src/compas_timber/elements/plate.py:60]: this takes the first point as frame.point and then the two adjacent points to make the frame.

File: src/compas_timber/ghpython/components/CT_Plate/code.py (hunk 4 above)
Comment (obucklin) [reply src/compas_timber/ghpython/components/CT_Plate/code.py:77]: ok I will leave it here for now

File: src/compas_timber/ghpython/components/CT_Model/code.py [hunk 6]
```diff
@@ -112,27 +112,37 @@ def get_joints_from_rules(self, beams, rules, topologies):
...
+        for element in Elements:
+            if element is None:
+                continue
+            element.remove_features()
+            element.debug_info = []
+            Model.add_element(element)
+            if element.__class__.__name__ == "Plate":
+                plates.append(element)
+            elif element.__class__.__name__ == "Beam":
+                beams.append(element)
```
Comment (chenkasirer) [inline src/compas_timber/ghpython/components/CT_Model/code.py:138]: I'd rather do:
`if isinstance(element, Plate)`

or better would be:
if element.is_plate

or even better than that would be:
we override `Model.add_element` and check it there

Comment (chenkasirer) [reply src/compas_timber/ghpython/components/CT_Model/code.py:138]:
```python3
    def add_element(self, element, **kwargs):
        node = super(TimberModel, self).add_element(element, **kwargs)

        if isinstance(element, Beam):
            self._beams.append(element)
        elif isinstance(element, Wall):
            self._walls.append(element)
        elif isinstance(element, Plate):
            self._plates.append(element)
        else:
            raise NotImplementedError("Element type not supported: {}".format(type(element)))
        return node
```

File: src/compas_timber/model/model.py [hunk 7]
```diff
@@ -45,18 +48,26 @@ def __from_data__(cls, data):
...
 
     @property
     def beams(self):
         # type: () -> list[Beam]
         return self._beams
 
+    @property
+    def plates(self):
+        # type: () -> list[Plate]
+        return self._plates
```
Comment (chenkasirer) [inline src/compas_timber/model/model.py:69]: `self._plates` should be declared in `TimberModel.__init__()`

File: src/compas_timber/ghpython/components/CT_Model/code.py [hunk 8]
```diff
@@ -162,18 +172,18 @@ def RunScript(self, Beams, JointRules, Features, MaxDistance, CreateGeometry):
...
             if CreateGeometry:
-                scene.add(beam.geometry)
-                if beam.debug_info:
-                    debug_info.add_feature_error(beam.debug_info)
+                scene.add(element.geometry)
+                if element.debug_info:
+                    debug_info.add_feature_error(element.debug_info)
             else:
-                scene.add(beam.blank)
+                scene.add(element.blank)
```
Comment (chenkasirer) [inline src/compas_timber/ghpython/components/CT_Model/code.py:186]: `if hasattr(element, "blank"): ...`

**Review by chenkasirer — APPROVED** [review]
LGTM!

**Pull request merged**